File: results/0.0/ELF/elf-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Define constants based on ELF specification
#define EI_NIDENT 16

// Parser for ELF Header
static HParser *elf_header() {
    return h_sequence(
        h_bits(32, false), // EI_MAG
        h_uint8(),  // EI_CLASS
        h_uint8(),  // EI_DATA
        h_uint8(),  // EI_VERSION
        h_uint8(),  // EI_OSABI
        h_uint8(),  // EI_ABIVERSION
        h_repeat_n(h_uint8(), 7), // EI_PAD
        h_uint16(), // e_type
        h_uint16(), // e_machine
        h_uint32(), // e_version
        h_uint64(), // e_entry
        h_uint64(), // e_phoff
        h_uint64(), // e_shoff
        h_uint32(), // e_flags
        h_uint16(), // e_ehsize
        h_uint16(), // e_phentsize
        h_uint16(), // e_phnum
        h_uint16(), // e_shentsize
        h_uint16(), // e_shnum
        h_uint16(), // e_shstrndx
        NULL
    );
}

// Parser for Section Header
static HParser *section_header() {
    return h_sequence(
        h_uint32(), // sh_name
        h_uint32(), // sh_type
        h_uint64(), // sh_flags
        h_uint64(), // sh_addr
        h_uint64(), // sh_offset
        h_uint64(), // sh_size
        h_uint32(), // sh_link
        h_uint32(), // sh_info
        h_uint64(), // sh_addralign
        h_uint64(), // sh_entsize
        NULL
    );
}

// Parser for Program Header
static HParser *program_header() {
    return h_sequence(
        h_uint32(), // p_type
        h_uint32(), // p_flags
        h_uint64(), // p_offset
        h_uint64(), // p_vaddr
        h_uint64(), // p_paddr
        h_uint64(), // p_filesz
        h_uint64(), // p_memsz
        h_uint64(), // p_align
        NULL
    );
}

// Main function to parse ELF file
int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <ELF file>\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp) {
        perror("Failed to open file");
        return 1;
    }

    // Seek to the end of the file to determine the size
    fseek(fp, 0, SEEK_END);
    size_t size = ftell(fp);
    fseek(fp, 0, SEEK_SET);

    // Read the entire file into memory
    uint8_t *buffer = malloc(size);
    if (!buffer) {
        perror("Failed to allocate memory");
        fclose(fp);
        return 1;
    }

    if (fread(buffer, 1, size, fp) != size) {
        perror("Failed to read file");
        free(buffer);
        fclose(fp);
        return 1;
    }

    fclose(fp);

    // Create a parser for an ELF file
    HParser *elf_parser = h_sequence(
        elf_header(),
        h_many(section_header()),
        h_many(program_header()),
        NULL
    );

    HParseResult *result = h_parse(elf_parser, buffer, size);
    if (result) {
        printf("ELF file parsed successfully.\n");
        h_pprint(stdout, result->ast, 0, 0);
    } else {
        fprintf(stderr, "Failed to parse ELF file.\n");
    }

    // Clean up
    h_parse_result_free(result);
    h_free_parser(elf_parser);
    free(buffer);

    return 0;
}

File: results/0.0/ELF/elf-gpt-4-turbo.rs
use nom::{
    bytes::complete::{take, tag},
    number::complete::{le_u16, le_u32, le_u64},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

#[derive(Debug)]
struct ElfHeader {
    e_ident: Vec<u8>,     // ELF Identification
    e_type: u16,          // Object file type
    e_machine: u16,       // Machine type
    e_version: u32,       // Object file version
    e_entry: u64,         // Entry point address
    e_phoff: u64,         // Program header offset
    e_shoff: u64,         // Section header offset
    e_flags: u32,         // Processor-specific flags
    e_ehsize: u16,        // ELF header size
    e_phentsize: u16,     // Size of program header entry
    e_phnum: u16,         // Number of program header entries
    e_shentsize: u16,     // Size of section header entry
    e_shnum: u16,         // Number of section header entries
    e_shstrndx: u16,      // Section header string table index
}

fn parse_elf_header(input: &[u8]) -> IResult<&[u8], ElfHeader> {
    let (input, e_ident) = tag(b"\x7FELF")(input)?;
    let (input, e_ident_rest) = take(12usize)(input)?;
    let mut e_ident_full = Vec::from(e_ident);
    e_ident_full.extend_from_slice(e_ident_rest);

    let (input, e_type) = le_u16(input)?;
    let (input, e_machine) = le_u16(input)?;
    let (input, e_version) = le_u32(input)?;
    let (input, e_entry) = le_u64(input)?;
    let (input, e_phoff) = le_u64(input)?;
    let (input, e_shoff) = le_u64(input)?;
    let (input, e_flags) = le_u32(input)?;
    let (input, e_ehsize) = le_u16(input)?;
    let (input, e_phentsize) = le_u16(input)?;
    let (input, e_phnum) = le_u16(input)?;
    let (input, e_shentsize) = le_u16(input)?;
    let (input, e_shnum) = le_u16(input)?;
    let (input, e_shstrndx) = le_u16(input)?;

    Ok((
        input,
        ElfHeader {
            e_ident: e_ident_full,
            e_type,
            e_machine,
            e_version,
            e_entry,
            e_phoff,
            e_shoff,
            e_flags,
            e_ehsize,
            e_phentsize,
            e_phnum,
            e_shentsize,
            e_shnum,
            e_shstrndx,
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <ELF file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_elf_header(&buffer) {
        Ok((_, elf_header)) => {
            println!("{:#?}", elf_header);
        }
        Err(e) => {
            eprintln!("Failed to parse ELF header: {:?}", e);
        }
    }

    Ok(())
}